kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt

```diff
@@ -311,19 +313,62 @@ class TunnelService : VpnService() {
         startForeground(STATUS_NOTIFICATION_ID, notification)
     }
 
+    private fun signInNotification() {
+        val chan = getNotificationChannel()
+        val notificationBuilder = NotificationCompat.Builder(this, chan.id)
+        val notification =
+            notificationBuilder
+                .setSmallIcon(R.drawable.ic_firezone_logo)
+                .setContentTitle(SIGN_IN_NOTIFICATION_TITLE)
+                .setContentText(SIGN_IN_NOTIFICATION_TEXT)
+                .setPriority(NotificationCompat.PRIORITY_HIGH)
+                .setCategory(NotificationCompat.CATEGORY_STATUS)
+                .setContentIntent(configIntent())
+                .setAutoCancel(true)
+                .build()
+
+        startForeground(SIGN_IN_NOTIFICATION_ID, notification)
+    }
+
+    private fun getNotificationChannel(): NotificationChannel {
+        if (notificationChannel == null) {
+            notificationChannel = NotificationChannel(
+                NOTIFICATION_CHANNEL_ID,
+                NOTIFICATION_CHANNEL_NAME,
+                NotificationManager.IMPORTANCE_DEFAULT,
+            ).apply { description = "firezone connection status" }
+            val manager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
+            manager.createNotificationChannel(notificationChannel!!)
+        }
+        return notificationChannel!!
+    }
+
+    private fun isSignedIn(): Boolean {
+        // See if there is a 'token' property set in managed configuration directly
+        val restrictionsManager = getSystemService(Context.RESTRICTIONS_SERVICE)
+                    as android.content.RestrictionsManager
+        val appRestrictions = restrictionsManager.applicationRestrictions
+        return !appRestrictions.containsKey(TOKEN_KEY)
+    }
+
     companion object {
         const val ACTION_CONNECT = "dev.firezone.android.tunnel.CONNECT"
         const val ACTION_DISCONNECT = "dev.firezone.android.tunnel.DISCONNECT"
 
         private const val NOTIFICATION_CHANNEL_ID = "firezone-connection-status"
         private const val NOTIFICATION_CHANNEL_NAME = "firezone-connection-status"
         private const val STATUS_NOTIFICATION_ID = 1337
-        private const val NOTIFICATION_TITLE = "Firezone Connection Status"
+        private const val STATUS_NOTIFICATION_TITLE = "Firezone Connection Status"
+        private const val SIGN_IN_NOTIFICATION_ID = 1338
+        private const val SIGN_IN_NOTIFICATION_TITLE = "Firezone Signed Out"
+        private const val SIGN_IN_NOTIFICATION_TEXT = "Sign in again"
 
         private const val TAG: String = "TunnelService"
         private const val SESSION_NAME: String = "Firezone Connection"
         private const val DEFAULT_MTU: Int = 1280
 
+        private const val TOKEN_KEY = "token"
+
```

Comment:
Here is where it can be hairy. We need to check if the user is signed in and not on a headless mode of operation. Usually the preference repository is responsible for this.

Should the `TunnelService` be dependent on the PreferenceRepository?

Reply:
One other way to think about it, is that the notification can just be a notification to the user that they are disconnected.

The connected status notification done by the `TunnelService` already sends the user to the `MainActivity` when they click it. This already prompts them to sign in again.

I could be overthinking it for now, so maybe down the road we'll have a better idea :)